components/builder-worker/src/vcs.rs

```diff
@@ -12,161 +12,80 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::path::Path;
+use std::path::{Path, PathBuf};
 
 use git2;
 use url::Url;
 
+use github_api_client::GithubApp;
+
 use error::{Error, Result};
 
 pub struct VCS {
     pub vcs_type: String,
     pub data: String,
-    pub auth_token: Option<String>,
-    pub username: Option<String>,
+    pub app_id: u64,
+    pub key_path: PathBuf,
+    pub installation_id: Option<u64>,
 }
 
 impl VCS {
     pub fn new(
         vcs_type: String,
         data: String,
-        auth_token: Option<String>,
-        username: Option<String>,
+        app_id: u64,
+        key_path: PathBuf,
+        installation_id: Option<u64>,
     ) -> VCS {
         VCS {
             vcs_type: vcs_type,
             data: data,
-            auth_token: auth_token,
-            username: username,
+            app_id: app_id,
+            installation_id: installation_id,
+            key_path: key_path,
         }
     }
 
     pub fn clone(&self, path: &Path) -> Result<()> {
         match self.vcs_type.as_ref() {
             "git" => {
+                let token = match self.installation_id {
+                    None => None,
+                    Some(id) => {
+                        let app = GithubApp::new(self.app_id, self.key_path.clone());
+                        Some(app.auth_as_installation(id).map_err(
+                            |e| Error::GithubAppAuthErr(e),
+                        )?)
+                    }
+                };
                 debug!(
                     "cloning git repository, url={}, path={:?}",
-                    self.url()?,
+                    self.url(token.clone())?,
                     path
                 );
-                git2::Repository::clone(&(self.url()?).as_str(), path)?;
+                git2::Repository::clone(&(self.url(token)?).as_str(), path)?;
                 Ok(())
             }
             _ => panic!("Unknown vcs type"),
         }
     }
 
-    pub fn url(&self) -> Result<Url> {
+    pub fn url(&self, token: Option<String>) -> Result<Url> {
```

Comment:
This can be `token: Option<&str>` if you call `token.as_ref()` on line 63. You're only using the ref on line 82 below so you don't need a fully allocated String